# 1.查看数据集整体

In [1]:
import os
files = os.listdir("ml-latest-small")

In [2]:
import re
datalist = []
for file in files:
    name = re.findall('(.*)\.(.*)',file)
    if name[0][1] == 'csv':
        datalist.append(file)
files = datalist

In [3]:
files
#一共有4个csv文件

['links.csv', 'tags.csv', 'ratings.csv', 'movies.csv']

# 2.分析4个数据集的具体信息

In [4]:
import pandas as pd

In [5]:
path = "ml-latest-small/{}".format(files[0])
links = pd.read_csv(path)

In [6]:
links.shape

(9742, 3)

In [7]:
links.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9742 entries, 0 to 9741
Data columns (total 3 columns):
movieId    9742 non-null int64
imdbId     9742 non-null int64
tmdbId     9734 non-null float64
dtypes: float64(1), int64(2)
memory usage: 228.4 KB


In [8]:
links.head(5)

movieId  imdbId   tmdbId
0        1  114709    862.0
1        2  113497   8844.0
2        3  113228  15602.0
3        4  114885  31357.0
4        5  113041  11862.0

**links.csv文件内容：**  
对于links.csv文件来说，一共是三列特征：   
* moviedId:为https://movielens.org 网站使用的电影的标识符  
* imdbld:为http://www.imdb.com 网站使用的电影标识符
* tmdbld:为https://www.themoviedb.org 网站使用的电影标识符号

以movieId=1为例
可以得到：  
在movielens中的网站信息为https://movielens.org/movies/1  
在imdbld中的网站信息为http://www.imdb.com/title/tt0114709/  
在tmdbld中的网站信息为https://www.themoviedb.org/movie/862  

**links.csv文件规模:**  
含有9742部电影的信息  
其中imdbid存在8个缺失值

**links.csv文件应用:**  
关于这个文件，目前想到的就是可以得到对应网站的相关介绍

In [9]:
path = "ml-latest-small/{}".format(files[1])
tags = pd.read_csv(path)

In [10]:
tags.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3683 entries, 0 to 3682
Data columns (total 4 columns):
userId       3683 non-null int64
movieId      3683 non-null int64
tag          3683 non-null object
timestamp    3683 non-null int64
dtypes: int64(3), object(1)
memory usage: 115.2+ KB


In [11]:
tags.shape

(3683, 4)

In [12]:
tags.head()

userId  movieId              tag   timestamp
0       2    60756            funny  1445714994
1       2    60756  Highly quotable  1445714996
2       2    60756     will ferrell  1445714992
3       2    89774     Boxing story  1445715207
4       2    89774              MMA  1445715200

In [13]:
#求打过标签的电影数目
len(tags.movieId.unique())

1572

**tags文件内容**  
包含4个特征  
* userId:用户id
* movieId:movielens中的id
* tag:用户打的元标签
* timestamp:打标签的时间戳

**tags文件规模**  
一共有3683条标签记录  
针对1572部电影打了标签，打标签的电影占电影总数的16%

**tags文件应用**   
可以用户打标签的习惯，比如倾向于对一部电影打几个标签  
提炼热门电影的关键词  
用于电影类型的分类  
衡量电影的热门程度  
打标签与打分数之间的关联

In [14]:
path = "ml-latest-small/{}".format(files[2])
ratings = pd.read_csv(path)

In [15]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100836 entries, 0 to 100835
Data columns (total 4 columns):
userId       100836 non-null int64
movieId      100836 non-null int64
rating       100836 non-null float64
timestamp    100836 non-null int64
dtypes: float64(1), int64(3)
memory usage: 3.1 MB


In [26]:
ratings.head()

userId  movieId  rating  timestamp
0       1        1     4.0  964982703
1       1        3     4.0  964981247
2       1        6     4.0  964982224
3       1       47     5.0  964983815
4       1       50     5.0  964982931

In [28]:
len(ratings.movieId.unique())

9724

**ratings文件内容**  
一共包含4个特征  
* userId:用户id
* moviedId:电影id
* rating:用户评分，评分以5星级为标准，以半星级递增
* timestamp:评分时间戳  

**ratings文件规模**  
一共包含100836条评分记录9742部电影的评分，有评分的电影占总电影数的99.8%

**ratings文件应用**  
可以基于用户的评分协同过滤进行电影推荐  
根据电影和类型评估哪类电影更容易得到高分  
评分时间是否和评分等级有关系  
衡量热门评分电影 
衡量高评分电影

In [17]:
path = "ml-latest-small/{}".format(files[3])
movies = pd.read_csv(path)

In [18]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9742 entries, 0 to 9741
Data columns (total 3 columns):
movieId    9742 non-null int64
title      9742 non-null object
genres     9742 non-null object
dtypes: int64(1), object(2)
memory usage: 228.4+ KB


In [19]:
movies.head(5)

movieId                               title  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance  
3                         Comedy|Drama|Romance  
4                                       Comedy

**movies文件内容**  
一共包含3个特征  
* movieId:电影在MovieLens网站上使用的电影ID
* title:电影名称，并且包含上映时间
* genres:电影类型  

**movies数据规模**  
共包含9742电影的信息,一共是20种类别，比数据说明文件中多IMAX这个类别
 
**movies数据应用**  
可以分析下每年出品的电影数目，以及细分到类型  
每种类型的电影的数目  
用于电影相关度的衡量  
电影类型的关联度，比如恐怖片往往是动作片  
IMAX的电影个数和电影比例

为了便于进行电影类型的分析，将genres拆开

In [20]:
movies_clean = pd.DataFrame()
for index in range(movies.shape[0]):
    line = movies.iloc[index,:]
    figures = line.genres
    figlist = re.findall("[\w -]+",figures)
    for figure in figlist:
        newrow = pd.DataFrame({"movieId":[line.movieId],"title":[line.title],"genres":[figure]})
        movies_clean = pd.concat([movies_clean,newrow])

In [21]:
movies_clean.head()

movieId             title     genres
0        1  Toy Story (1995)  Adventure
0        1  Toy Story (1995)  Animation
0        1  Toy Story (1995)   Children
0        1  Toy Story (1995)     Comedy
0        1  Toy Story (1995)    Fantasy

In [23]:
movies_clean.genres.unique()

array(['Adventure', 'Animation', 'Children', 'Comedy', 'Fantasy',
       'Romance', 'Drama', 'Action', 'Crime', 'Thriller', 'Horror',
       'Mystery', 'Sci-Fi', 'War', 'Musical', 'Documentary', 'IMAX',
       'Western', 'Film-Noir', 'no genres listed'], dtype=object)

# 3.将清洗后的数据进行保存

In [169]:
links.to_csv('cleaned_data/links.csv',index=False)

movies_clean.to_csv('cleaned_data/movies.csv',index=False)

ratings.to_csv('cleaned_data/ratings.csv',index=False)

tags.to_csv('cleaned_data/tags.csv',index=False)